### ISSUE TITLE
SeekableStream Enhancements

### ISSUE THREAD

**Issue by DanRStevens** [OutpostUniverse/OP2Utility#24]
Seekable streams should have a well defined position which can be both read and written. It also makes sense they would have a well defined length, as otherwise you could not bounds check a seek.

A common feature in seeking is to support both absolute and relative offsets. Some APIs support this through an extra parameter to the seek, though in the vast majority of cases, such a parameter ends up being a constant. I think a better design is to provide two separate methohds Seek and SeekRelative.

Further, large streams, such as files exceeding 4GB should be supported, even on 32-bit builds. As such, stream positions should be at minimum 64 bits.

As such, I propose SeekableStream should have the following methods:

```
	// Get the size of the stream
	virtual uint64_t Length() = 0;
	// Get the current position, from the start of the stream
	virtual uint64_t Position() = 0;
	// Seek to absolute position, given as offset from beginning of stream
	virtual void Seek(uint64_t position) = 0;
	// Seek ahead by a relative amount, given as offset from current position
	virtual void SeekRelative(uint64_t offset) = 0;
```

**Comment by Brett208** [comment]
What do you think about allowing the user to seek backwards relative by making SeekRelative a int64_t?

I noticed std::fstream allows seeking backwards. Although I'm currently struggling to come up with a reason I would want to seek backwards. (I guess you could still seek backwards using Seek instead of SeekRelative in this construct.)

Is this an enhancement you are looking to tackle or having me work on?

Thanks.
-Brett

**Comment by DanRStevens** [comment]
Seeking backwards makes sense. Can be useful if you peek at data to make a decision, such as reading a header marker to determine file or data type, and selecting an appropriate parser. The parser would most likely try to read and verify the marker itself.

Seeking backwards is more intuitive with an `int64_t` as opposed to a `uint64_t`, though unsigned values do have well defined wrap around, so it would be possible (and identical) either way. Signed values have less well defined wrap around. Though I could go either way for SeekRelative.

I can probably work on this enhancement. Though I noticed my use of Seek and SeekRelative is opposite of yours. If I just blindly implemented as suggested, it would break existing code.

**Comment by Brett208** [comment]
I have no problem with what I am calling Seek becoming SeekRelative. We could push an update to the codebase renaming Seek to SeekRelative before you start, or I'm also happy if you just rename them all as you go.

For testing purposes, I have been using OP2MapImager. Seeing if it can render the map is a pretty good test of the MapReader/StreamReader code. Inside OP2MapImager, I have locally inserted calls to Write the map to disk after the map is rendered. Then I will toss the saved map into Outpost 2 and see if it works. Of course there are probably a lot of other ways to test. OP2MapImager does not match with the current version of OP2Utility. I'll look at updating it shortly though.

**Comment by Brett208** [comment]
I was realizing that fstream::seekg takes a type called std::streamoff

> It is a typdef of one the fundamental signed integral types large enough to represent the maximum possible file size supported by the system.
> It is the type returned by subtracting two stream position of an fpos type, and can be converted to and from this type.

I'm not sure if that will take a uint64_t or not? I guess it should based on the definition? If not we may have to stop using it for these changes. Just wanted to point it out.

**Comment by DanRStevens** [comment]
The issue with using `std::streamoff` in a parameter, is it creates a leaky abstraction that depends on `ios`. If you want to do this, you should create a new opaque data type specific to your stream library, which translates to that underlying type.

Most likely, `streamoff` will be a `uint64_t` or compatible type. If it was smaller, say `uint32_t`, it would only be able to address 4GB of data, which is not big enough for some files, such as DVD images. This was a common limitation of many file utilities about 10 years ago, and usually not because the file API lacked the width, but simply because internal program code stored offsets in `int` datatypes, which were only 32-bits wide. I'm not aware of filesystems that take larger than 64-bit addresses (16 Exbibytes).

**Comment by Brett208** [comment]
@DanRStevens,

If you are planning on using fstream in the improved SeekableStreamEnhancements, I stumbled across this article on how to implement the Length function: https://stackoverflow.com/questions/22984956/tellg-function-give-wrong-size-of-file/22986486#22986486

Which would roughly equate to below. I don't really know what to do with the streamsize type though. Perhaps casting it to size_t is a mistake?

```
size_t FileStreamReader::Length() {
	file.ignore(std::numeric_limits<std::streamsize>::max());
	std::streamsize length = file.gcount();
	file.clear();   //  Since ignore will have set eof.
	file.seekg(0, std::ios_base::beg);

	return length;
}
```
-Brett

**Comment by DanRStevens** [comment]
Yes, it is unnecessarily messy. The C/C++ API can really show its age.

I noticed the article said the length discrepancy was only an issue for text mode files. Probably to do with EOL conversions, and using two character codes on Windows (CR LF), but only counting it as a single return character. Using binary mode avoids this issue.

Reading the documentation for `gcount`, it returns the number of characters extracted by the last formatted input. The file will need to be at the beginning for that method to return the full file size. If any reading is done first, or if the file is opened in append mode, or if any seeking is done, it will instead return the number of bytes left. It also unconditionally restores the stream position to the beginning. That might be unexpected if someone tried to determine file size during the middle of file processing. The method is destructive to the internal stream state.

I had implemented the method (ignoring text mode issues) in my earlier attempt as:
```
uint64_t StreamFileReader::length() {
	auto currentPosition = file.tellg();	// Record current position
	file.seekg(0, std::ios::end);			// Seek to end of file
	auto length = file.tellg();				// Record current position (length of file)
	file.seekg(currentPosition);			// Restore position
	return length;
}
```
